packages/utils/props.ts

```diff
@@ -33,44 +54,69 @@ export function buildProp<
   type,
   validator,
 }: {
-  values?: readonly T[]
+  type?: T
+  values?: readonly V[]
   required?: R
   default?: R extends true
     ? never
     : D extends Record<string, unknown> | Array<any>
     ? () => D
     : D
-  type?: any
   validator?: ((val: any) => val is C) | ((val: any) => boolean)
 } = {}) {
-  type DefaultType = typeof defaultValue
-  type HasDefaultValue = Exclude<T, D> extends never ? false : true
+  type HasDefaultValue = Exclude<D, undefined> extends never ? false : true
+  type Type = PropType<
+    | (T extends PropWrapper<unknown>
+        ? T[typeof wrapperKey]
+        : [V] extends [never]
+        ? ResolvePropTypeWithReadonly<T>
+        : never)
+    | V
+    | C
+  >
 
   return {
-    type: type as PropType<T | C>,
+    type: ((type as any)?.[wrapperKey] || type) as unknown as Type,
     required: !!required as R,
 
-    default: defaultValue as R extends true
+    default: defaultValue as unknown as R extends true
       ? never
       : HasDefaultValue extends true
-      ? Exclude<DefaultType, undefined>
+      ? Exclude<
+          D extends Record<string, unknown> | Array<any> ? () => D : D,
+          undefined
+        >
       : undefined,
 
     validator:
       values || validator
         ? (val: unknown) => {
             let valid = false
-            if (values)
-              valid ||= ([...values, defaultValue] as unknown[]).includes(val)
+            if (values) {
+              const allowValues: unknown[] = [...values, defaultValue]
+              valid ||= allowValues.includes(val)
+              if (!valid) {
+                debugWarn(
+                  `Vue warn`,
+                  `Invalid prop: Expected one of (${allowValues.join(
+                    ', '
+                  )}), got value ${val}`
+                )
```

Comment:
I think this part might be a little bit redundant since Vue itself will warn the user when validator returns false.

Reply:
Vue will only warn that the validate is not passed, but the specific acceptable value is not informed.